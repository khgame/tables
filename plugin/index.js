
module.exports = {
  rows: require('./rows'),
  mark: require('./mark'),
  desc: require('./desc'),
  expand: require('./expand'),
  plain: require('./plain')
}
