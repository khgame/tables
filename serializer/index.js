const {
  readAndTranslate
} = require('../read')
const Path = require('path')
const fs = require('fs-extra')

/**
 * serialize files with selected serializers
 * @param {string} pathIn - input file path
 * @param {string} dirOut - output dir
 * @param {Object.<string,Serializer>} serializers
 */
function serialize (pathIn, dirOut, serializers) {
  let plugins = Object.values(serializers).reduce((prev, cur) => prev.concat(cur.plugins), [])
  // console.log('plugins', plugins)
  // console.dir(Object.values(serializers))
  let ret = readAndTranslate(pathIn, {
    plugins
  })
  for (let outName in serializers) {
    let fileContent = serializers[outName].file(ret, Path.parse(pathIn).name)
    fs.outputFileSync(Path.resolve(dirOut, outName), fileContent)
  }
}

module.exports = {
  serialize,
  jsonSerializer: require('./jsonSerializer').jsonSerializer,
  tsInterfaceSerializer: require('./tsInterfaceSerializer').tsInterfaceSerializer
}
