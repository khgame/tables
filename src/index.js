export * from './plugin'
export * from './utils/read'
export * from './utils/names'
export * from './serializer'
