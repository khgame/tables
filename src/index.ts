export * from './plugin'
export * from './utils/read'
export * from './utils/names'
export * from './serializer'
export * from './core/protocol'
export * from './pipeline/layers'

export { exportTablesToTs } from './outOfBox/exportTablesToTs'
