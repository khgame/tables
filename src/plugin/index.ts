export * from './rows'
export * from './erows'
export * from './colMap'
export * from './mark'
export * from './desc'
export * from './expand'
export * from './plain'
export * from './convert'
export * from './schema'

