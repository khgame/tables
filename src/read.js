const XLSX = require('js-xlsx')

function readWorkBook (path) {
  // read the raw file
  return XLSX.readFile(path)
}

function translateWorkBook (workbook, sheetName) {
  // read the sheet with the given name or the first sheet
  const targetSheetInd = workbook.SheetNames.findIndex(n => n === (sheetName || '__data'))
  const targetSheetName = workbook.SheetNames[targetSheetInd > 0 ? targetSheetInd : 0]
  const sheet = workbook.Sheets[targetSheetName]

  // create return table
  let data = {}
  let colKeys = []

  // convert data types
  for (let key in sheet) {
    if (!sheet.hasOwnProperty(key)) continue
    let splitInd = -1
    for (let i = key.length - 1; i >= 0; i--) {
      if (key[i] >= '0' && key[i] <= '9') {
        continue
      }
      splitInd = i
      break
    }
    let col = key.slice(0, splitInd + 1)
    let row = key.slice(splitInd + 1)
    if (!row) { continue }
    if (!data[row]) data[row] = {}
    let { t, v, w } = sheet[key]
    data[row][col] = { t, v, w }
    colKeys[col] = 1
  }

  let cols = Object.keys(colKeys).sort((a, b) => a.length - b.length || a.localeCompare(b)) // 1. length 2. locale
  return {
    cols,
    data,
    getValue: (table_, row_, col_) => (!table_.data[row_]) || (!table_.data[row_][col_]) ? undefined : table_.data[row_][col_].v
  }
}

function readAndTranslate (path, options) {
  let table = translateWorkBook(readWorkBook(path), options.sheetName)
  if (options.plugins) {
    table = options.plugins.reduce((accumulate, plugin) => plugin(accumulate), table)
  }
  return table
}

module.exports = {
  readWorkBook,
  translateWorkBook,
  readAndTranslate
}
