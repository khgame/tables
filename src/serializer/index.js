const {
  readAndTranslate
} = require('../utils/read')
const Path = require('path')
const fs = require('fs-extra')

/**
 * serialize files with selected serializers
 * @param {string} pathIn - input file path
 * @param {string} dirOut - output dir
 * @param {Object.<string,Serializer>} serializers
 */
function serialize (pathIn, dirOut, serializers) {
  let plugins = Object.values(serializers).reduce((prev, cur) => prev.concat(cur.plugins), [])
  // console.log('plugins', plugins)
  // console.dir(Object.values(serializers))
  let ret = readAndTranslate(pathIn, {
    plugins
  })
  for (let outName in serializers) {
    let fileContent = serializers[outName].file(ret, Path.parse(pathIn).name)
    fs.outputFileSync(Path.resolve(dirOut, outName), fileContent)
  }
}

module.exports = {
  serialize,
  jsonSerializer: require('./jsonSerializer').jsonSerializer,
  jsSerializer: require('./jsSerializer').jsSerializer,
  tsSerializer: require('./tsSerializer').tsSerializer,
  tsInterfaceSerializer: require('./tsInterfaceSerializer').tsInterfaceSerializer
}
