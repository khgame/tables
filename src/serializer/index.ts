export * from './core'
export * from './formats/json'
export * from './formats/js'
export * from './formats/ts'
export * from './formats/tsInterface'
export * from './formats/jsonx'
export * from './formats/go'
export * from './formats/csharp'
