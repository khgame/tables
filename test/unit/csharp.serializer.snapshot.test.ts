import { readAndTranslate } from '../../src/utils/read'
import { tableSchema, tableConvert } from '../../src/plugin'
import { csharpSerializer } from '../../src/serializer'
import * as Path from 'path'

describe('csharp serializer snapshot (example.xlsx)', () => {
  it('matches C# class snapshot', () => {
    const excel = Path.resolve(__dirname, '../../example/example.xlsx')
    const data = readAndTranslate(excel, { plugins: [tableSchema, tableConvert] })
    const code = csharpSerializer.file(data as any, 'example', '', {})
    expect(code).toMatchSnapshot()
  })
})
