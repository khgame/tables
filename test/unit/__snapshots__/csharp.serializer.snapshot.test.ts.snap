// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`csharp serializer snapshot (example.xlsx) matches C# class snapshot 1`] = `
"// Code generated by tables. DO NOT EDIT.
#nullable enable
using System;
using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace Tables
{
    public class Example
    {
    [JsonPropertyName("ctype")]
        public double Ctype { get; set; }

    [JsonPropertyName("building")]
        public double Building { get; set; }

    [JsonPropertyName("level")]
        public double Level { get; set; }

    [JsonPropertyName("name")]
        public string Name { get; set; }

    [JsonPropertyName("upgrage")]
        public ExampleUpgrage Upgrage { get; set; }

    [JsonPropertyName("product")]
        public List<ExampleProductItem> Product { get; set; }

    [JsonPropertyName("cost")]
        public List<ExampleCostItem> Cost { get; set; }

    [JsonPropertyName("arr")]
        public List<double> Arr { get; set; }

    [JsonPropertyName("pair")]
        public ExamplePair Pair { get; set; }

    [JsonPropertyName("map")]
        public List<ExampleMapItem> Map { get; set; }

    [JsonPropertyName("nest")]
        public List<object> Nest { get; set; }

    [JsonPropertyName("stars")]
        public List<object> Stars { get; set; }

    [JsonPropertyName("nestedArray")]
        public List<object> NestedArray { get; set; }

    [JsonPropertyName("ax")]
        public object Ax { get; set; }
     }

    public class ExampleUpgrage
    {
    [JsonPropertyName("to")]
        public double To { get; set; }

    [JsonPropertyName("dependency")]
        public List<double> Dependency { get; set; }
     }

    public class ExampleProductItem
    {
    [JsonPropertyName("tid")]
        public double Tid { get; set; }

    [JsonPropertyName("num")]
        public double Num { get; set; }
     }

    public class ExampleCostItem
    {
    [JsonPropertyName("key")]
        public string Key { get; set; }

    [JsonPropertyName("val")]
        public double Val { get; set; }
     }

    public class ExamplePair
    {
    [JsonPropertyName("key")]
        public string Key { get; set; }

    [JsonPropertyName("val")]
        public double Val { get; set; }
     }

    public class ExampleMapItem
    {
    [JsonPropertyName("key")]
        public string Key { get; set; }

    [JsonPropertyName("val")]
        public object Val { get; set; }
     }

    public class ExampleNestedArrayItem0
    {
    [JsonPropertyName("data")]
        public double Data { get; set; }
     }

    public class ExampleNestedArrayItem1
    {
    [JsonPropertyName("data")]
        public double Data { get; set; }
     }
}"
`;
